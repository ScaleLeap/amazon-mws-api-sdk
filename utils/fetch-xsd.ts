import { writeFileSync, mkdirSync } from 'fs'
import { join, dirname } from 'path'
import { fetch } from 'cross-fetch'

// https://stackoverflow.com/questions/8931175/where-are-the-specifications-xsds-for-amazon-mws-feed-xml-processing-reports

// https://images-na.ssl-images-amazon.com/images/G/01/rainier/help/xsd/release_1_9/OrderReport.xsd

const BASE_DIR = join(__dirname, '../xsd/')
const BASE_URL = 'https://images-na.ssl-images-amazon.com/images/G/01/rainier/help/xsd/release_4_1/'
const SCHEMA_LOCATION_REGEX = new RegExp('<xsd:include schemaLocation="([^"]+)"', 'gm')
const fetched = new Map()

async function download(filename: string) {
  const res = await fetch(`${BASE_URL}/${filename}`)

  return await res.text()
}

function extractSchemaLocations(xml: string) {
  return [...xml.matchAll(SCHEMA_LOCATION_REGEX)].map((match) => match[1])
}

async function get(filename: string): Promise<boolean> {
  console.debug('Getting file: %s', filename)

  const filePath = join(BASE_DIR, filename)

  if (fetched.get(filePath)) {
    console.debug('>> File %s already fecthed. Skipping...')
    return false
  }

  const res = await download(filename)
  const dir = dirname(filename)

  if (dir !== '.') {
    mkdirSync(join(BASE_DIR, dir), {
      recursive: true,
    })
  }

  writeFileSync(filePath, res, { encoding: 'utf8' })

  fetched.set(filePath, true)

  const extractedSchemaLocations = extractSchemaLocations(res)

  if (extractedSchemaLocations.length > 0) {
    await Promise.all(extractedSchemaLocations.map((f) => get(f)))
  }

  return true
}

async function main() {
  return get('amzn-envelope.xsd')
}

main()
