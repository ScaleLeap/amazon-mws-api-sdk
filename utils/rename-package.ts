import { readFileSync, writeFileSync } from 'fs'
import { join } from 'path'

import { name } from '../package.json'

const NEW_NAME = 'mws'

function changeName(file: string) {
  const path = join(__dirname, '../', file)
  let content = readFileSync(path, { encoding: 'utf8' })
  content = content.split(name).join(NEW_NAME)
  writeFileSync(path, content, { encoding: 'utf8' })
}

changeName('package.json')
changeName('package-lock.json')
changeName('README.md')
