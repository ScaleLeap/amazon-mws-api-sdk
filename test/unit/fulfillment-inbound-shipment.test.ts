import {
  CreateInboundShipmentParameters,
  InboundShipmentHeader,
  ParsingError,
  PartneredSmallParcelPackageInput,
} from '../../src'
import { createMockHttpClient, mockMwsFail, mockMwsServiceStatus, parsingError } from '../utils'

function mockFunctions() {
  /**
   * Mock everything in purify-ts, restore it back to normal,
   *  except for `enumeration` which will be stubbed to accept
   * https://github.com/facebook/jest/issues/936#issuecomment-265074320
   */
  const original = jest.requireActual('purify-ts')

  return {
    ...original, // Pass down all the exported objects
    enumeration: <T extends Record<string, string | number>>(enumeration: T) => {
      const enumValues = Object.values(enumeration)

      return original.Codec.custom({
        decode: (input: string | number) => {
          if (typeof input !== 'string' && typeof input !== 'number') {
            return original.Left(`Expected enum, received ${input}`)
          }

          const enumIndex = enumValues.indexOf(input)

          return enumIndex !== -1 || input === 'String'
            ? original.Right((enumValues[enumIndex] as T[keyof T]) || 'String')
            : original.Left(`Expected enum, received ${input}`)
        },
        encode: original.identity,
        schema: () => ({ enum: enumValues }),
      })
    },
  }
}
jest.mock('purify-ts', () => mockFunctions())

const mockAddress = {
  Name: '',
  AddressLine1: '',
  Email: '',
  City: '',
  PostalCode: '',
  CountryCode: '',
  Phone: '',
}

const mockInboundShipmentItem = {
  SellerSKU: '',
  QuantityShipped: 1,
}

const mockInboundShipmentHeader: InboundShipmentHeader = {
  ShipmentName: '',
  ShipFromAddress: mockAddress,
  DestinationFulfillmentCenterId: '',
  LabelPrepPreference: 'SELLER_LABEL',
  ShipmentStatus: 'WORKING',
}

describe('fulfillmentInboundShipment', () => {
  describe('getTransportContent', () => {
    const parameters = {
      ShipmentId: ''
    }

    it('returns the transport content if succesful', async () => {
      expect.assertions(1)

      const mockGetTransportContent = createMockHttpClient(
        'fulfillment_inbound_shipment_get_transport_content',
      )

      expect(
        await mockGetTransportContent.fulfillmentInboundShipment.getTransportContent(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response  is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getTransportContent(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('estimateTransportRequest', () => {
    const parameters = {
      ShipmentId: '',
    }

    it('returns the workflow status for a shipment if succesful', async () => {
      expect.assertions(1)

      const mockEstimateTransportRequest = createMockHttpClient(
        'fulfillment_inbound_shipment_estimate_transport_request',
      )

      expect(
        await mockEstimateTransportRequest.fulfillmentInboundShipment.estimateTransportRequest(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response i snt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.estimateTransportRequest(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('putTransportContent', () => {
    const mockPartneredSmallParcelPackageInput: PartneredSmallParcelPackageInput = {
      Dimensions: {
        Unit: 'inches',
        Length: 1,
        Width: 1,
        Height: 1,
      },
      Weight: {
        Unit: 'pounds',
        Value: 1,
      },
    }

    const mockTransportDetailInput = {
      PartneredSmallParcelData: {
        CarrierName: '',
        PackageList: [mockPartneredSmallParcelPackageInput],
      },
    }

    const parameters = {
      ShipmentId: '',
      IsPartnered: true,
      ShipmentType: 'SP',
      TransportDetails: mockTransportDetailInput,
    }

    it('returns tranport result if succesful', async () => {
      expect.assertions(1)

      const mockPutTransportContent = createMockHttpClient(
        'fulfillment_inbound_shipment_put_transport_content',
      )

      expect(
        await mockPutTransportContent.fulfillmentInboundShipment.putTransportContent(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response i snt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.putTransportContent(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getPrepInstructionsForAsin', () => {
    const parameters = {
      ASINList: [''],
      ShipToCountryCode: 'US',
    }

    it('returns list of prep instructions if succesful', async () => {
      expect.assertions(1)

      const mockGetPrepInstructionsForAsin = createMockHttpClient(
        'fulfillment_inbound_shipment_get_prep_instructions_for_asin',
      )

      expect(
        await mockGetPrepInstructionsForAsin.fulfillmentInboundShipment.getPrepInstructionsForAsin(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is nt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getPrepInstructionsForAsin(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getPrepInstructionsForSku', () => {
    const parameters = {
      SellerSKUList: [],
      ShipToCountryCode: 'US',
    }

    it('returns prep instructions list if succesful', async () => {
      expect.assertions(1)

      const mockGetPrepInstructionsForSku = createMockHttpClient(
        'fulfillment_inbound_shipment_get_prep_instructions_for_sku',
      )

      expect(
        await mockGetPrepInstructionsForSku.fulfillmentInboundShipment.getPrepInstructionsForSku(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is nt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getPrepInstructionsForSku(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('confirmPreorder', () => {
    const parameters = {
      ShipmentId: '',
      NeedByDate: new Date(),
    }

    it('returns confirmed dates if succesful', async () => {
      expect.assertions(1)

      const mockConfirmPreorder = createMockHttpClient(
        'fulfillment_inbound_shipment_confirm_preorder',
      )

      expect(
        await mockConfirmPreorder.fulfillmentInboundShipment.confirmPreorder(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response isn t valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.confirmPreorder(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getPreorderInfo', () => {
    const parameters = { ShipmentId: '' }

    it('returns preorderinfo if succesful', async () => {
      expect.assertions(1)

      const mockGetPreorderInfo = createMockHttpClient(
        'fulfillment_inbound_shipment_get_preorder_info',
      )

      expect(
        await mockGetPreorderInfo.fulfillmentInboundShipment.getPreorderInfo(parameters),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response isn t valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getPreorderInfo(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('updateInboundShipment', () => {
    const parameters = {
      ShipmentId: '',
      InboundShipmentHeader: mockInboundShipmentHeader,
      InboundShipmentItems: [mockInboundShipmentItem],
    }

    it('return the shipment id if succesful', async () => {
      expect.assertions(1)

      const mockUpdateInboundShipment = createMockHttpClient(
        'fulfillment_inbound_shipment_update_inbound_shipment',
      )

      expect(
        await mockUpdateInboundShipment.fulfillmentInboundShipment.updateInboundShipment(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it("throws a parsing error when the status response isn't valid", async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.updateInboundShipment(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('createInboundShipment', () => {
    const parameters: CreateInboundShipmentParameters = {
      ShipmentId: '',
      InboundShipmentHeader: mockInboundShipmentHeader,
      InboundShipmentItems: [mockInboundShipmentItem],
    }

    it('returns the shipment ID if succesful', async () => {
      expect.assertions(1)

      const mockCreateInboundShipment = createMockHttpClient(
        'fulfillment_inbound_shipment_create_inbound_shipment',
      )

      expect(
        await mockCreateInboundShipment.fulfillmentInboundShipment.createInboundShipment(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it("throws a parsing error when the status response isn't valid", async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.createInboundShipment(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('createInboundShipmentPlan', () => {
    const mockInboundShipmentPlanRequestItem = {
      SellerSKU: '',
      Quantity: 1,
    }

    const parameters = {
      ShipFromAddress: mockAddress,
      InboundShipmentPlanRequestItems: [mockInboundShipmentPlanRequestItem],
    }

    it('returns inbound shipment plans if succesful', async () => {
      expect.assertions(1)

      const mockCreateInboundShipmentPlan = createMockHttpClient(
        'fulfillment_inbound_shipment_create_inbound_shipment_plan',
      )

      expect(
        await mockCreateInboundShipmentPlan.fulfillmentInboundShipment.createInboundShipmentPlan(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response isnt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.createInboundShipmentPlan(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getInboundGuidanceForAsin', () => {
    const parameters = {
      ASINList: [],
      MarketplaceId: '',
    }

    it('returns sku inbound guidance list if succesful', async () => {
      expect.assertions(1)

      const mockGetInboundGuidanceForSku = createMockHttpClient(
        'fulfillment_inbound_shipment_get_inbound_guidance_for_asin',
      )

      expect(
        await mockGetInboundGuidanceForSku.fulfillmentInboundShipment.getInboundGuidanceForAsin(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response isnt valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getInboundGuidanceForAsin(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getInboundGuidanceForSku', () => {
    const parameters = {
      SellerSKUList: [],
      MarketplaceId: '',
    }

    it('returns sku inbound guidance list if succesful', async () => {
      expect.assertions(1)

      const mockGetInboundGuidanceForSku = createMockHttpClient(
        'fulfillment_inbound_shipment_get_inbound_guidance_for_sku',
      )

      expect(
        await mockGetInboundGuidanceForSku.fulfillmentInboundShipment.getInboundGuidanceForSku(
          parameters,
        ),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getInboundGuidanceForSku(parameters),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })

  describe('getServiceStatus', () => {
    it('returns a parsed model when the status response is valid', async () => {
      expect.assertions(1)

      expect(
        await mockMwsServiceStatus.fulfillmentInboundShipment.getServiceStatus(),
      ).toMatchSnapshot()
    })

    it('throws a parsing error when the status response is not valid', async () => {
      expect.assertions(1)

      await expect(() =>
        mockMwsFail.fulfillmentInboundShipment.getServiceStatus(),
      ).rejects.toStrictEqual(new ParsingError(parsingError))
    })
  })
})
