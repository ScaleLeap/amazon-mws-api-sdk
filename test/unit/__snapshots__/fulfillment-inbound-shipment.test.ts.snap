// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fulfillmentInboundShipment createInboundShipment returns the shipment ID if succesful 1`] = `
Array [
  Object {
    "ShipmentId": "FBA2F50GYX",
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`fulfillmentInboundShipment createInboundShipmentPlan returns inbound shipment plans if succesful 1`] = `
Array [
  Object {
    "InboundShipmentPlans": Array [
      Object {
        "DestinationFulfillmentCenterId": "String",
        "EstimatedBoxContentsFee": Object {
          "FeePerUnit": Object {
            "CurrencyCode": "String",
            "Value": "100",
          },
          "TotalFee": Object {
            "CurrencyCode": "String",
            "Value": "100",
          },
          "TotalUnits": 1,
        },
        "Items": Array [
          Object {
            "FulfillmentNetworkSKU": "String",
            "PrepDetailsList": Array [
              Object {
                "PrepInstruction": "String",
                "PrepOwner": "String",
              },
            ],
            "Quantity": 1,
            "SellerSKU": "String",
          },
        ],
        "LabelPrepType": "String",
        "ShipToAddress": Object {
          "AddressLine1": "String",
          "AddressLine2": "String",
          "City": "String",
          "CountryCode": "String",
          "DistrictOrCounty": "String",
          "Name": "String",
          "PostalCode": "String",
          "StateOrProvinceCode": "String",
        },
        "ShipmentId": "String",
      },
    ],
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`fulfillmentInboundShipment getInboundGuidanceForAsin returns sku inbound guidance list if succesful 1`] = `
Array [
  Object {
    "ASINInboundGuidanceList": Array [
      Object {
        "ASIN": "B00005N5PF",
        "GuidanceReasonList": Array [
          "SlowMovingASIN",
        ],
        "InboundGuidance": "InboundNotRecommended",
      },
    ],
    "InvalidASINList": Array [],
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`fulfillmentInboundShipment getInboundGuidanceForSku returns sku inbound guidance list if succesful 1`] = `
Array [
  Object {
    "InvalidSKUList": Array [],
    "SKUInboundGuidanceList": Array [
      Object {
        "ASIN": "B00005N5PF",
        "GuidanceReasonList": Array [
          "SlowMovingASIN",
        ],
        "InboundGuidance": "InboundNotRecommended",
        "SellerSKU": "us001",
      },
    ],
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`fulfillmentInboundShipment getServiceStatus returns a parsed model when the status response is valid 1`] = `
Array [
  Object {
    "Status": "GREEN",
    "Timestamp": "2020-05-06T08:22:23.582Z",
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;

exports[`fulfillmentInboundShipment updateInboundShipment return the shipment id if succesful 1`] = `
Array [
  Object {
    "ShipmentId": "FBA2F50GYX",
  },
  Object {
    "quotaMax": 1000,
    "quotaRemaining": 999,
    "quotaResetOn": 2020-04-06T10:22:23.582Z,
    "requestId": "0",
    "timestamp": 2020-05-06T09:22:23.582Z,
  },
]
`;
